File: avocado/commands/ext/dnf.py
from code import interact
from sys import intern
"""Extension dnf command implementation."""

import sys
from avocado.commands.base import BaseCommand
from avocado.utils.container import SdkContainer
from avocado.utils.config import load_config
from avocado.utils.output import print_error, print_info, print_success
from avocado.utils.target import resolve_target, get_target_from_config


class ExtDnfCommand(BaseCommand):
    """Implementation of the 'ext dnf' command."""

    def _do_extension_setup(
        self, config, extension, container_helper, container_image, target, verbose, container_args=None
    ):
        """Perform extension setup for DNF operations."""
        # Check if extension directory structure exists, create it if not
        check_cmd = f"test -d $AVOCADO_EXT_SYSROOTS/extensions/{extension}"
        repo_url = config.get("sdk", {}).get("repo_url")
        repo_release = config.get("sdk", {}).get("repo_release")

        print_info(repo_release)
        dir_exists = container_helper.run_in_container(
            container_image=container_image,
            target=target,
            command=check_cmd,
            verbose=False,  # Don't show verbose output for the check
            source_environment=False,
            repo_url=repo_url,
            repo_release=repo_release,
            container_args=container_args,
        )
        if not dir_exists:
            print_info(f"Creating sysroot for extension '{extension}'.")
            setup_commands = [
                f"mkdir -p $AVOCADO_EXT_SYSROOTS/extensions/{
                    extension}/var/lib",
                f"cp -rf $AVOCADO_EXT_SYSROOTS/rootfs/var/lib/rpm $AVOCADO_EXT_SYSROOTS/extensions/{
                    extension}/var/lib",
            ]

            setup_success = container_helper.run_in_container(
                container_image=container_image,
                target=target,
                command=setup_commands,
                verbose=verbose,
                source_environment=False,
                repo_url=repo_url,
                repo_release=repo_release,
                container_args=container_args,
            )

            if setup_success:
                print_success(f"Created sysroot for extension '{extension}'.")
            else:
                print_error(
                    f"Failed to create sysroot for extension '{extension}'.")
                return False

        return True

    @classmethod
    def register_subparser(cls, subparsers):
        """Register the ext dnf command's subparser."""
        parser = subparsers.add_parser(
            "dnf", help="Run DNF commands in an extension's context"
        )

        # Add common arguments
        parser.add_argument(
            "-C",
            "--config",
            default="avocado.toml",
            help="Path to avocado.toml configuration file (default: avocado.toml)",
        )

        parser.add_argument(
            "-v", "--verbose", action="store_true", help="Enable verbose output"
        )

        parser.add_argument(
            "--container-args",
            help="Additional arguments to pass to the container runtime"
        )

        # Extension is now required as a flag
        parser.add_argument(
            "-e",
            "--extension",
            required=True,
            help="Extension name to operate on"
        )

        # Command is now required as a flag
        parser.add_argument(
            "-c",
            "--command",
            nargs="*",
            required=True,
            help="DNF command and arguments to execute",
        )

        return parser

    def execute(self, args, parser=None, unknown=None):
        """Execute the ext dnf command."""
        # Get command args from -c/--command flag
        command_args = getattr(args, "command", [])
        if unknown:
            command_args.extend(unknown)

        if not command_args:
            print(
                "Error: No DNF command specified.",
                file=sys.stderr,
            )
            if parser:
                parser.print_help()
            return False

        # Parse container args if provided
        container_args = None
        if args.container_args:
            import shlex
            container_args = shlex.split(args.container_args)

        config_path = args.config
        extension = args.extension

        # Load configuration
        config, success = load_config(config_path)
        if not success:
            return False

        # Get the SDK image from configuration
        container_image = config.get("sdk", {}).get("image")
        if not container_image:
            print(
                "Error: No container image specified in config under 'sdk.image'",
                file=sys.stderr,
            )
            return False

        # Get repo_url from config, if it exists
        repo_url = config.get("sdk", {}).get("repo_url")
        repo_release = config.get("sdk", {}).get("repo_release")

        # Use resolved target (from CLI/env) if available, otherwise fall back to config
        config_target = get_target_from_config(config)
        target = resolve_target(
            cli_target=args.resolved_target, config_target=config_target
        )
        if not target:
            print(
                "Error: No target architecture specified. Use --target, AVOCADO_TARGET env var, or config under 'runtime.<name>.target'.",
                file=sys.stderr,
            )
            return False

        container_helper = SdkContainer()

        # Check if extension exists in configuration
        if "ext" not in config or extension not in config["ext"]:
            print_error(
                f"Extension '{
                    extension}' not found in configuration."
            )
            return False

        # Do extension setup first
        if not self._do_extension_setup(
            config, extension, container_helper, container_image, target, args.verbose, container_args
        ):
            return False

        # Build dnf command with extension installroot
        installroot = f"$AVOCADO_EXT_SYSROOTS/{extension}"
        dnf_cmd = f"$DNF_SDK_HOST --installroot={
            installroot} {' '.join(command_args)}"

        # Run the DNF command using the container helper
        success = container_helper.run_in_container(
            container_image=container_image,
            target=target,
            command=dnf_cmd,
            verbose=args.verbose,
            source_environment=False,
            use_entrypoint=True,
            repo_url=repo_url,
            repo_release=repo_release,
            container_args=container_args,
        )

        # Log the result
        if success:
            print_success(f"DNF command completed successfully.")
        else:
            print_error(f"DNF command failed.")

        return success

    def _print_help(self):
        """Print custom help message."""
        print(
            "usage: avocado ext dnf [-h] [-C CONFIG] [-v] -e EXTENSION -c COMMAND...")
        print()
        print("Execute DNF commands in an extension's context")
        print()
        print("options:")
        print("  -h, --help            show this help message and exit")
        print("  -C CONFIG, --config CONFIG")
        print(
            "                        Path to avocado.toml configuration file (default: avocado.toml)"
        )
        print("  -v, --verbose         Enable verbose output")
        print("  -e EXTENSION, --extension EXTENSION")
        print("                        Extension name to operate on (required)")
        print("  -c COMMAND, --command COMMAND")
        print("                        DNF command and arguments to execute (required)")
        print("  --container-args CONTAINER_ARGS")
        print(
            "                        Additional arguments to pass to the container runtime")
        print()
        print("Examples:")
        print("  avocado ext dnf -e myext -c repolist --enabled")
        print("  avocado ext dnf -e myext -c search python")
        print("  avocado ext dnf -e myext -c install vim")
        print("  avocado ext dnf -e myext -c --reinstall install gcc")
        print("  avocado ext dnf -e myext -c list installed")
        print("  avocado ext dnf --container-args '--network user-avocado' -e myext -c repolist")

File: src/commands/runtime/dnf.rs
use anyhow::{Context, Result};
use std::sync::Arc;

use crate::utils::config::{ComposedConfig, Config};
use crate::utils::container::{RunConfig, SdkContainer};
use crate::utils::output::{print_error, print_info, print_success, OutputLevel};
use crate::utils::target::resolve_target_required;

pub struct RuntimeDnfCommand {
    config_path: String,
    runtime: String,
    command: Vec<String>,
    verbose: bool,
    target: Option<String>,
    container_args: Option<Vec<String>>,
    dnf_args: Option<Vec<String>>,
    sdk_arch: Option<String>,
    /// Pre-composed configuration to avoid reloading
    composed_config: Option<Arc<ComposedConfig>>,
}

impl RuntimeDnfCommand {
    pub fn new(
        config_path: String,
        runtime: String,
        command: Vec<String>,
        verbose: bool,
        target: Option<String>,
        container_args: Option<Vec<String>>,
        dnf_args: Option<Vec<String>>,
    ) -> Self {
        Self {
            config_path,
            runtime,
            command,
            verbose,
            target,
            container_args,
            dnf_args,
            sdk_arch: None,
            composed_config: None,
        }
    }

    /// Set SDK container architecture for cross-arch emulation
    pub fn with_sdk_arch(mut self, sdk_arch: Option<String>) -> Self {
        self.sdk_arch = sdk_arch;
        self
    }

    /// Set pre-composed configuration to avoid reloading
    #[allow(dead_code)]
    pub fn with_composed_config(mut self, config: Arc<ComposedConfig>) -> Self {
        self.composed_config = Some(config);
        self
    }

    pub async fn execute(&self) -> Result<()> {
        // Use provided config or load fresh
        let composed = match &self.composed_config {
            Some(cc) => Arc::clone(cc),
            None => Arc::new(
                Config::load_composed(&self.config_path, self.target.as_deref())
                    .context("Failed to load composed config")?,
            ),
        };
        let config = &composed.config;
        let merged_container_args = config.merge_sdk_container_args(self.container_args.as_ref());
        let parsed = &composed.merged_value;

        self.validate_runtime_exists(parsed)?;
        let container_image = self.get_container_image(config)?;
        let target = self.resolve_target_architecture(config)?;

        // Get repo_url and repo_release from config
        let repo_url = config.get_sdk_repo_url();
        let repo_release = config.get_sdk_repo_release();

        self.execute_dnf_command(
            parsed,
            &container_image,
            &target,
            repo_url.as_ref(),
            repo_release.as_ref(),
            &merged_container_args,
        )
        .await
    }

    fn validate_runtime_exists(&self, parsed: &serde_yaml::Value) -> Result<()> {
        let runtime_section = parsed.get("runtimes").ok_or_else(|| {
            print_error(
                &format!("Runtime '{}' not found in configuration.", self.runtime),
                OutputLevel::Normal,
            );
            anyhow::anyhow!("No runtime section found")
        })?;

        let runtime_table = runtime_section
            .as_mapping()
            .ok_or_else(|| anyhow::anyhow!("Invalid runtime section format"))?;

        if !runtime_table.contains_key(&self.runtime) {
            print_error(
                &format!("Runtime '{}' not found in configuration.", self.runtime),
                OutputLevel::Normal,
            );
            return Err(anyhow::anyhow!("Runtime not found"));
        }

        Ok(())
    }

    fn get_container_image(&self, config: &Config) -> Result<String> {
        config
            .get_sdk_image()
            .map(|s| s.to_string())
            .ok_or_else(|| {
                anyhow::anyhow!("No container image specified in config under 'sdk.image'.")
            })
    }

    fn resolve_target_architecture(&self, config: &crate::utils::config::Config) -> Result<String> {
        resolve_target_required(self.target.as_deref(), config)
    }

    #[allow(clippy::too_many_arguments)]
    async fn execute_dnf_command(
        &self,
        parsed: &serde_yaml::Value,
        container_image: &str,
        target: &str,
        repo_url: Option<&String>,
        repo_release: Option<&String>,
        merged_container_args: &Option<Vec<String>>,
    ) -> Result<()> {
        let container_helper = SdkContainer::new();

        // Perform runtime setup first
        self.setup_runtime_environment(
            parsed,
            &container_helper,
            container_image,
            target,
            repo_url,
            repo_release,
            merged_container_args,
        )
        .await?;

        // Build and execute DNF command
        let dnf_command = self.build_dnf_command();
        self.run_dnf_command(
            &container_helper,
            container_image,
            target,
            &dnf_command,
            repo_url,
            repo_release,
            merged_container_args,
        )
        .await
    }

    #[allow(clippy::too_many_arguments)]
    async fn setup_runtime_environment(
        &self,
        _config: &serde_yaml::Value,
        container_helper: &SdkContainer,
        container_image: &str,
        target: &str,
        repo_url: Option<&String>,
        repo_release: Option<&String>,
        merged_container_args: &Option<Vec<String>>,
    ) -> Result<()> {
        let check_cmd = format!("test -d $AVOCADO_PREFIX/runtimes/{}", self.runtime);

        let config = RunConfig {
            container_image: container_image.to_string(),
            target: target.to_string(),
            command: check_cmd,
            verbose: self.verbose,
            source_environment: false, // don't source environment
            interactive: false,
            repo_url: repo_url.cloned(),
            repo_release: repo_release.cloned(),
            container_args: merged_container_args.clone(),
            dnf_args: self.dnf_args.clone(),
            sdk_arch: self.sdk_arch.clone(),
            ..Default::default()
        };
        let dir_exists = container_helper.run_in_container(config).await?;

        if !dir_exists {
            self.create_runtime_directory(
                container_helper,
                container_image,
                target,
                repo_url,
                repo_release,
                merged_container_args,
            )
            .await?;
        }

        Ok(())
    }

    #[allow(clippy::too_many_arguments)]
    async fn create_runtime_directory(
        &self,
        container_helper: &SdkContainer,
        container_image: &str,
        target: &str,
        repo_url: Option<&String>,
        repo_release: Option<&String>,
        merged_container_args: &Option<Vec<String>>,
    ) -> Result<()> {
        let setup_cmd = format!(
            "mkdir -p $AVOCADO_PREFIX/runtimes/{}/var/lib && cp -rf $AVOCADO_PREFIX/rootfs/var/lib/rpm $AVOCADO_PREFIX/runtimes/{}/var/lib",
            self.runtime, self.runtime
        );

        let config = RunConfig {
            container_image: container_image.to_string(),
            target: target.to_string(),
            command: setup_cmd,
            verbose: self.verbose,
            source_environment: false, // don't source environment
            interactive: false,
            repo_url: repo_url.cloned(),
            repo_release: repo_release.cloned(),
            container_args: merged_container_args.clone(),
            dnf_args: self.dnf_args.clone(),
            sdk_arch: self.sdk_arch.clone(),
            ..Default::default()
        };
        let setup_success = container_helper.run_in_container(config).await?;

        if !setup_success {
            print_error(
                &format!("Failed to set up runtime directory for '{}'.", self.runtime),
                OutputLevel::Normal,
            );
            return Err(anyhow::anyhow!("Failed to create runtime directory"));
        }

        if self.verbose {
            print_info(
                &format!("Created runtime directory for '{}'.", self.runtime),
                OutputLevel::Normal,
            );
        }

        Ok(())
    }

    #[allow(clippy::too_many_arguments)]
    async fn run_dnf_command(
        &self,
        container_helper: &SdkContainer,
        container_image: &str,
        target: &str,
        dnf_command: &str,
        repo_url: Option<&String>,
        repo_release: Option<&String>,
        merged_container_args: &Option<Vec<String>>,
    ) -> Result<()> {
        if self.verbose {
            print_info(
                &format!("Running DNF command: {dnf_command}"),
                OutputLevel::Normal,
            );
        }

        let config = RunConfig {
            container_image: container_image.to_string(),
            target: target.to_string(),
            command: dnf_command.to_string(),
            verbose: self.verbose,
            source_environment: true, // source environment for DNF
            interactive: true,
            repo_url: repo_url.cloned(),
            repo_release: repo_release.cloned(),
            container_args: merged_container_args.clone(),
            dnf_args: self.dnf_args.clone(),
            sdk_arch: self.sdk_arch.clone(),
            ..Default::default()
        };
        let success = container_helper.run_in_container(config).await?;

        if success {
            print_success("DNF command completed successfully.", OutputLevel::Normal);
            Ok(())
        } else {
            print_error("DNF command failed.", OutputLevel::Normal);
            Err(anyhow::anyhow!("DNF command failed"))
        }
    }

    fn build_dnf_command(&self) -> String {
        let installroot = format!("$AVOCADO_PREFIX/runtimes/{}", self.runtime);
        let command_args_str = self.command.join(" ");
        let dnf_args_str = if let Some(args) = &self.dnf_args {
            format!(" {} ", args.join(" "))
        } else {
            String::new()
        };

        format!(
            r#"
RPM_ETCCONFIGDIR="$DNF_SDK_TARGET_PREFIX" \
$DNF_SDK_HOST \
    $DNF_NO_SCRIPTS \
    $DNF_SDK_TARGET_REPO_CONF \
    --setopt=sslcacert=${{SSL_CERT_FILE}} \
    --installroot={installroot} \
    --disablerepo=${{AVOCADO_TARGET}}-target-ext \
    {dnf_args_str} \
    {command_args_str}
"#
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_new() {
        let cmd = RuntimeDnfCommand::new(
            "avocado.yaml".to_string(),
            "test-runtime".to_string(),
            vec!["list".to_string(), "installed".to_string()],
            false,
            Some("x86_64".to_string()),
            None,
            None,
        );

        assert_eq!(cmd.config_path, "avocado.yaml");
        assert_eq!(cmd.runtime, "test-runtime");
        assert_eq!(cmd.command, vec!["list", "installed"]);
        assert!(!cmd.verbose);
        assert_eq!(cmd.target, Some("x86_64".to_string()));
    }

    #[test]
    fn test_new_with_verbose_and_args() {
        let cmd = RuntimeDnfCommand::new(
            "avocado.yaml".to_string(),
            "test-runtime".to_string(),
            vec!["install".to_string(), "gcc".to_string()],
            true,
            None,
            Some(vec!["--cap-add=SYS_ADMIN".to_string()]),
            Some(vec!["--nogpgcheck".to_string()]),
        );

        assert_eq!(cmd.config_path, "avocado.yaml");
        assert_eq!(cmd.runtime, "test-runtime");
        assert_eq!(cmd.command, vec!["install", "gcc"]);
        assert!(cmd.verbose);
        assert_eq!(cmd.target, None);
        assert_eq!(
            cmd.container_args,
            Some(vec!["--cap-add=SYS_ADMIN".to_string()])
        );
        assert_eq!(cmd.dnf_args, Some(vec!["--nogpgcheck".to_string()]));
    }

    #[test]
    fn test_build_dnf_command() {
        let cmd = RuntimeDnfCommand::new(
            "avocado.yaml".to_string(),
            "test-runtime".to_string(),
            vec!["list".to_string(), "installed".to_string()],
            false,
            None,
            None,
            Some(vec!["--nogpgcheck".to_string()]),
        );

        let dnf_command = cmd.build_dnf_command();

        assert!(dnf_command.contains("--installroot=$AVOCADO_PREFIX/runtimes/test-runtime"));
        assert!(dnf_command.contains("list installed"));
        assert!(dnf_command.contains("--nogpgcheck"));
        assert!(dnf_command.contains("RPM_ETCCONFIGDIR"));
        assert!(dnf_command.contains("$DNF_SDK_HOST"));
    }

    #[test]
    fn test_build_dnf_command_no_args() {
        let cmd = RuntimeDnfCommand::new(
            "avocado.yaml".to_string(),
            "my-runtime".to_string(),
            vec!["search".to_string(), "python".to_string()],
            false,
            None,
            None,
            None,
        );

        let dnf_command = cmd.build_dnf_command();

        assert!(dnf_command.contains("--installroot=$AVOCADO_PREFIX/runtimes/my-runtime"));
        assert!(dnf_command.contains("search python"));
        assert!(!dnf_command.contains("--nogpgcheck"));
    }
}
